src/antelope/wallets/authenticators/EVMAuthenticator.ts

```diff
@@ -0,0 +1,81 @@
+// EVMAuthenticator class
+
+import { SendTransactionResult } from '@wagmi/core';
+import { BigNumber, ethers } from 'ethers';
+import { useChainStore } from 'src/antelope/stores/chain';
+import { useEVMStore } from 'src/antelope/stores/evm';
+import { createTraceFunction, isTracingAll, useFeedbackStore } from 'src/antelope/stores/feedback';
+import { AntelopeError, EvmTransactionResponse, ExceptionError, TokenClass, addressString } from 'src/antelope/types';
+
+export abstract class EVMAuthenticator {
+    readonly label: string;
+    readonly trace: (message: string, ...args: unknown[]) => void;
+
+    constructor(label: string) {
+        this.label = label;
+        const name = `${this.getName()}(${label})`;
+        this.trace = createTraceFunction(name);
+        useFeedbackStore().setDebug(name, isTracingAll());
+    }
+    abstract getName(): string;
+    abstract logout(): Promise<void>;
+    abstract getSystemTokenBalance(address: addressString | string): Promise<BigNumber>;
+    abstract getERC20TokenBalance(address: addressString | string, tokenAddress: addressString | string): Promise<BigNumber>;
+    abstract transferTokens(token: TokenClass, amount: BigNumber, to: addressString | string): Promise<EvmTransactionResponse | SendTransactionResult>;
+    abstract prepareTokenForTransfer(token: TokenClass | null, amount: BigNumber, to: string): Promise<void>;
+    abstract isConnectedTo(chainId: string): Promise<boolean>;
+    abstract externalProvider(): Promise<ethers.providers.ExternalProvider>;
+    abstract web3Provider(): Promise<ethers.providers.Web3Provider>;
+
+    // to easely clone the authenticator
+    abstract newInstance(label: string): EVMAuthenticator;
+
+    // la función login debería ser posible reimplementar por subclases de esta clase abtracta
```

Comment:
can you make this in english just for consistency?

Reply:
Perdón !!! :see_no_evil: